### PR TITLE
pr921 Define cmsDriver step HGCalTriggerPrimitives

### PR DESCRIPTION
**Pull request by rekovic**
pr921 for PhaseII L1T MC processing.

Define step HGCalTriggerPrimitives to be used by cmsDriver command for PhaseII L1T MC reprocessing.

Reason:  
Since PhaseII MC samples were made, there was an important development in 
HGCalTriggerPrimitives available in CMSSW_9_2_1.
To be able to reprocess the samples centrally and run only the new HGCalTPs, need to define
it as a step to be used by `cmsDriver.py`

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by kpedro88** [comment]
assign upgrade

File: Configuration/Applications/python/ConfigBuilder.py [hunk 1]
```diff
@@ -1602,6 +1604,12 @@ def prepare_L1TrackTrigger(self, sequence = "L1TrackTrigger"):
 	self.scheduleSequence(sequence.split('.')[-1],'L1TrackTrigger_step')
         return
 
+    def prepare_HGCalTriggerPrimitives(self, sequence = "HGCalTriggerPrimitives"):
+        ''' Enrich the schedule with L1 reconstruction '''
+        self.loadDefaultOrSpecifiedCFF(sequence,self.HGCalTriggerPrimitivesDefaultCFF)
+	self.scheduleSequence(sequence.split('.')[-1],'HGCalTriggerPrimitives_step')
```
Comment (kpedro88) [inline Configuration/Applications/python/ConfigBuilder.py]: please use consistent indentation in python code (don't mix tabs and spaces)

**Comment by rekovic** [comment]
@kpedro88, @davidlange6  
As discussed at the ORP, due to changed geometry, might have to go back to 90X,
in which case we would need 90x backport of HGCal TPs PR https://github.com/cms-sw/cmssw/pull/18861.

**Comment by kpedro88** [comment]
@jbsauvan FYI

Does #18861 work for both HGCal geometries (v7 and v8)?

**Comment by rekovic** [comment]
please test

**Comment by jbsauvan** [comment]
There are still missing pieces for the V8 geometry, so the HGCAL TPG won't work with V8. I'm trying to get  everything working with V8 this week.

**Comment by kpedro88** [comment]
okay, that's actually better for this purpose - L1 TDR samples were generated with 2023D4, which has v7 HGCal.

**Comment by rekovic** [comment]
@kpedro88 
From the response of @jbsauvan, I was to understand that we were OK to use it on v7 geometry.

- In CMSSW_920 + rebased #18861 + rebased this PR,
I tried to run this step `-s HGCalTriggerPrimitives` on L1T MC samples and it runs fine.
Example command:
`cmsDriver.py step2 --python_filename=test_rerun_step2_HGCalTPs_onMCL1_FEVTHLTDEBUG.py --no_exec -s HGCalTriggerPrimitives --datatier GEN-SIM-DIGI-RAW -n 5 --era Phase2C2 --eventcontent FEVTDEBUGHLT --pileup AVE_200_BX_25ns --filein root://cms-xrd-global.cern.ch///store/mc/PhaseIISpring17D/SingleNeutrino/GEN-SIM-DIGI-RAW/PU200_90X_upgrade2023_realistic_v9-v1/70000/004C6E8B-5F26-E711-AE46-0242AC130002.root --conditions 90X_upgrade2023_realistic_v1 --era Phase2C2_timing --beamspot HLLHC14TeV --geometry Extended2023D4 --fileout file:step2_PU200_rerun-HGCalTPs.root
`

- In CMSSW_921 + this PR, the same command above fails
<pre>
----- Begin Fatal Exception 07-Jun-2017 10:36:38 CEST-----------------------
An exception of category 'NoProxyException' occurred while
   [0] Processing  stream begin Run run: 1 stream: 0
   [1] Calling method for module HGCalTriggerDigiProducer/'hgcalTriggerPrimitiveDigiProducer'
   [2] Using EventSetup component HGCalTriggerGeometryESProducer/'hgcalTriggerGeometryESProducer' to make data HGCalTriggerGeometryBase/'' in record IdealGeometryRecord
Exception Message:
No data of type "HGCalTopology" with label "HGCalEESensitive" in record "IdealGeometryRecord"
 Please add an ESSource or ESProducer to your job which can deliver this data.
----- End Fatal Exception -------------------------------------------------
</pre>

Addtiional note:  need to add all products "HGCalTriggerPrimitives" to EventContent FEVTDEBUGHLT.

**Comment by rekovic** [comment]
please test

**Comment by kpedro88** [comment]
@rekovic The 2023D4 geometry no longer exists in 9_2_1 - onward. This is why the PRs need to be backported to 90X for compatibility with the L1 TDR samples.

**Comment by davidlange6** [comment]
in any case the usual model for trigger primitives is that they are defined as part of other steps (remembering that you can use a syntax like L1:hgcalTriggerPrimitives to run an alternative sequence) . putting these changes into ConfigBuilder is not the right place for them.

**Comment by rekovic** [comment]
@davidlange6 
>@rekovic The 2023D4 geometry no longer exists in 9_2_1 - onward. This is why the PRs need to be backported to 90X for compatibility with the L1 TDR samples.

Since this runs fine on top of 9_2_0, a patch to 9_2_0 would be fine?
PRs (this one + rebased #18861) need to be made to which branch?  9_2_0_patchX ?

**Comment by kpedro88** [comment]
@rekovic there have been numerous other changes since 9_0_X that could cause inconsistencies...

**Comment by davidlange6** [comment]
we need a master PR and then a 90x PR it seems.

**Comment by rekovic** [comment]
@kpedro88 @davidlange6 
OK.  Understood.  Need PRs to 90x.  Thanks.

**Comment by rekovic** [comment]
@davidlange6 
>in any case the usual model for trigger primitives is that they are defined as part of other steps (remembering that you can use a syntax like L1:hgcalTriggerPrimitives to run an alternative sequence) . putting these changes into ConfigBuilder is not the right place for them.

Ok.  Added hgcalTriggerPrimitives to phase2_ecalDigiSequence, and reverted the changes to ConfigBulider.  This step `-s DIGI:hgcalTriggerPrimitives` runs fine and correct products are saved, in CMSSW_9_2_0 with command

`cmsDriver.py step2 --python_filename=test_rerun_step2_HGCalTPs_onMCL1_FEVTHLTDEBUG.py --no_exec -s DIGI: --datatier GEN-SIM-DIGI-RAW -n 5 --era Phase2C2 --eventcontent FEVTDEBUGHLT --pileup AVE_200_BX_25ns --filein root://cms-xrd-global.cern.ch///store/mc/PhaseIISpring17D/SingleNeutrino/GEN-SIM-DIGI-RAW/PU200_90X_upgrade2023_realistic_v9-v1/70000/004C6E8B-5F26-E711-AE46-0242AC130002.root --conditions 90X_upgrade2023_realistic_v1 --era Phase2C2_timing --beamspot HLLHC14TeV --geometry Extended2023D4 --fileout file:step2_PU200_rerun-HGCalTPs.root
`

**Comment by rekovic** [comment]
please test

**Comment by kpedro88** [comment]
1) `ecalDigiSequence` is not the right place for `hgcalTriggerPrimitives`. These are entirely separate subdetectors.
2) Rather than reverting the previous commits and leaving a messy history, just rebase them out of the branch entirely.

**Comment by rekovic** [comment]
@kpedro88 
>1. `ecalDigiSequence` is not the right place for `hgcalTriggerPrimitives`. These are entirely separate subdetectors.

>2. Rather than reverting the previous commits and leaving a messy history, just rebase them out of the branch entirely.
1) You are right of course.  A proper thing to do is to add `SimCalorimetry/Configuration/python/hgcalDigiSequence_cff.py`.  will do.
2) Indeed, can make a PR with clean history once arrived to proper solution.

**Comment by kpedro88** [comment]
@rekovic what is the status of this PR?

**Pull request closed without merging**